internal/collector/sysinfo/hardware/hardware_linux.go

```diff
@@ -424,6 +446,43 @@ func (h Collector) collectScreens(pi platform.Info) (info []screen, err error) {
 			RefreshRate: v[2],
 		})
 	}
-
 	return info, nil
 }
+
+type realWaylandProvider struct{}
+
+func (*realWaylandProvider) InitWayland() int {
+	return int(C.init_wayland())
+}
+
+func (h Collector) cScreensWayland() (screens []screen, err error) {
+	waylandMutex.Lock()
+	defer waylandMutex.Unlock()
+	if h.platform.wayland.InitWayland() != 0 {
+		return nil, fmt.Errorf("failed to connect to Wayland display")
+	}
+	defer C.cleanup()
+
+	if bool(C.had_memory_error()) {
+		h.log.Warn("Memory error while trying to get displays")
+		return nil, fmt.Errorf("failed to get displays")
+	}
+
+	count := int(C.get_output_count())
+	screens = make([]screen, count)
+	displays := C.get_displays()
+	for i := range count {
+		display := *(**C.struct_wayland_display)(unsafe.Pointer(uintptr(unsafe.Pointer(displays)) + uintptr(i)*unsafe.Sizeof((*C.struct_wayland_display)(nil))))
```

Comment:
This written in the middle of some Go code looks quite ugly... but I couldn't find a nicer way that wouldn't be too verbose for little benefit.